torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -682,176 +712,139 @@ Stmt* GPUMetaVarRewriter::mutate(const For* v) {
 
     const Var* metaVar = gpu_thread_vars_[gpu_thread_index];
     body = Substitute(Stmt::clone(body), {{v->var(), metaVar}});
+  }
 
-    // extents is the max found, if we don't equal that we may be smaller.
-    // Mask it with the current loops stop.
-    if (!exprEquals(extents[gpu_thread_index], v->stop())) {
-      body = new Cond(
-          new CompareSelect(
-              metaVar,
-              v->stop(),
-              new IntImm(1),
-              new IntImm(0),
-              CompareSelectOperation::kLT),
-          body,
-          nullptr);
-    }
+  // Recurse into body block.
+  body = Stmt::clone(body->accept_mutator(this));
+
+  // pop the internal rank off the stack.
+  if (loop_options.is_gpu_block_index()) {
+    current_block_rank_[loop_options.gpu_block_index()] = old_rank;
+    return body;
+  } else if (loop_options.is_gpu_thread_index()) {
+    current_thread_rank_[loop_options.gpu_thread_index()] = old_rank;
     return body;
   }
 
-  return new For(
-      v->var(), v->start(), v->stop(), Stmt::clone(body), loop_options);
+  return v->cloneWithNewBody(body);
 }
 
 Stmt* GPUMetaVarRewriter::mutate(const Block* v) {
-  bool any_change = false;
+  std::vector<Segment> innerSegments;
+  Segment current;
 
-  std::vector<Stmt*> stmts;
+  auto pushAndReset = [&](bool unmask) {
+    if (!current.empty()) {
+      innerSegments.push_back(current);
+    }
+    current.reset(unmask);
+  };
+
+  // Here's we're slicing the Block's contents into segments that should have
+  // the same launch rank. Segments are comprised of all statements that aren't
+  // loops - which are their own segments. Some operations, such as threading
+  // and memory ops should never be masked and so also get their own segment.
   for (Stmt* stmt : *v) {
     Stmt* stmt_new = stmt->accept_mutator(this);
-    if (stmt != stmt_new) {
-      any_change = true;
-    } else {
-      stmt_new = Stmt::clone(stmt);
+    if (stmt == stmt_new) {
+      stmt_new = Stmt::clone(stmt_new);
     }
 
     if (need_sync_) {
-      stmts.push_back(new SyncThreads());
+      // sync is special, we never want to mask it and it is never part of
+      // another segment.
+      pushAndReset(true);
+      current.stmts.push_back(new SyncThreads());
```

Comment:
Minor: this is fine for me. But syncthreads can be part of the mask, as long as all threads go through the same branch. But that is a corner case that we can ignore for now.

Reply:
Yeah, but then we'd need analysis to know that all threads went through the branch - this way it should *just work*.